ccl/changefeedccl: TestChangefeedNemeses failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/f97dc13163020a032b098ef3eb88e4d9f54a04ba

Parameters:
```
TAGS=
GOFLAGS=-parallel=4
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestChangefeedNemeses PKG=github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1614352&tab=buildLog

```

panic: rangefeed processor unexpectedly stopped

goroutine 117608 [running]:
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).setRangefeedProcessor(0xc00070c800, 0xc0004b94a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_rangefeed.go:260 +0x127
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).registerWithRangefeedRaftMuLocked(0xc00070c800, 0x46b18c0, 0xc0021b6a50, 0xc0003c70f0, 0x3, 0x8, 0xc0021873f0, 0x2, 0x2, 0x15daf31046d12eff, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_rangefeed.go:380 +0x56f
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).RangeFeed(0xc00070c800, 0xc002ffe780, 0x46f6980, 0xc000735ac0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_rangefeed.go:229 +0x41f
github.com/cockroachdb/cockroach/pkg/storage.(*Store).RangeFeed(0xc0011d0700, 0xc002ffe780, 0x46f6980, 0xc000735ac0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:2196 +0x116
github.com/cockroachdb/cockroach/pkg/storage.(*Stores).RangeFeed(0xc002a0d040, 0xc002ffe780, 0x46f6980, 0xc000735ac0, 0xc001991f70)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/stores.go:206 +0x12b
github.com/cockroachdb/cockroach/pkg/server.(*Node).RangeFeed(0xc000625400, 0xc002ffe780, 0x46f6980, 0xc000735ac0, 0xc001991f78, 0x18a8f4a)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:1042 +0x59
github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.RangeFeed.func1(0xc002746560, 0x466d080, 0xc000625400, 0xc002ffe780, 0x46b1800, 0xc001a348c0, 0xc00186f500, 0xc00186f560)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:604 +0xbe
created by github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.RangeFeed
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:602 +0x11d

goroutine 1 [chan receive]:
testing.(*T).Run(0xc00258c500, 0x3d92fb9, 0x15, 0x3ee6568, 0x7ae201)
	/usr/local/go/src/testing/testing.go:917 +0x381
testing.runTests.func1(0xc000215d00)
	/usr/local/go/src/testing/testing.go:1157 +0x78
testing.tRunner(0xc000215d00, 0xc000757e10)
	/usr/local/go/src/testing/testing.go:865 +0xc0
testing.runTests(0xc000318e60, 0x6bd6900, 0x31, 0x31, 0x0)
	/usr/local/go/src/testing/testing.go:1155 +0x2a9
testing.(*M).Run(0xc000203980, 0x0)
	/usr/local/go/src/testing/testing.go:1072 +0x162
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestMain(0xc000203980)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/main_test.go:31 +0xda
main.main()
	_testmain.go:138 +0x13e

goroutine 11 [syscall, 1 minutes]:
os/signal.signal_recv(0x0)
	/usr/local/go/src/runtime/sigqueue.go:139 +0x9c
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:23 +0x22
created by os/signal.init.0
	/usr/local/go/src/os/signal/signal_unix.go:29 +0x41

goroutine 12 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.flushDaemon()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:73 +0x74
created by github.com/cockroachdb/cockroach/pkg/util/log.init.7
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:45 +0x35

goroutine 13 [chan receive, 1 minutes]:
github.com/cockroachdb/cockroach/pkg/util/log.signalFlusher()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:100 +0x126
created by github.com/cockroachdb/cockroach/pkg/util/log.init.7
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:46 +0x4d

goroutine 15 [select]:
github.com/cockroachdb/cockroach/vendor/go.opencensus.io/stats/view.(*worker).start(0xc000300600)
	/go/src/github.com/cockroachdb/cockroach/vendor/go.opencensus.io/stats/view/worker.go:152 +0xdd
created by github.com/cockroachdb/cockroach/vendor/go.opencensus.io/stats/view.init.0
	/go/src/github.com/cockroachdb/cockroach/vendor/go.opencensus.io/stats/view/worker.go:29 +0x57

goroutine 114364 [select]:
database/sql.(*DB).connectionResetter(0xc002856600, 0x46b1800, 0xc0029bfe00)
	/usr/local/go/src/database/sql/sql.go:1013 +0xfb
created by database/sql.OpenDB
	/usr/local/go/src/database/sql/sql.go:671 +0x194

goroutine 113760 [select]:
```

--- a/pkg/storage/allocator_scorer.go
+++ b/pkg/storage/allocator_scorer.go
@@ -640,8 +640,8 @@ func rebalanceCandidates(
 
 	// 3. Decide whether we should try to rebalance. Note that for each existing
 	// store, we only compare its fullness stats to the stats of "comparable"
-	// stores, i.e. those stores that at least as valid, necessary, and diverse
-	// as the existing store.
+	// stores, i.e. those stores that are at least as valid, necessary, and
+	// diverse as the existing store.
 	needRebalance := needRebalanceFrom || needRebalanceTo
 	var shouldRebalanceCheck bool
 	if !needRebalance {

--- a/pkg/storage/rangefeed/processor.go
+++ b/pkg/storage/rangefeed/processor.go
@@ -210,6 +210,9 @@ func (p *Processor) Start(stopper *stop.Stopper, rtsIter engine.SimpleIterator) 
 				// Add the new registration to the registry.
 				p.reg.Register(&r)
 
+				// Publish an updated filter that includes the new registration.
+				p.filterResC <- p.reg.NewFilter()
+
 				// Immediately publish a checkpoint event to the registry. This will be
 				// the first event published to this registration after its initial
 				// catch-up scan completes.
@@ -346,7 +349,9 @@ func (p *Processor) sendStop(pErr *roachpb.Error) {
 // engine which occurred after the provided start timestamp.
 //
 // If the method returns false, the processor will have been stopped, so calling
-// Stop is not necessary.
+// Stop is not necessary. If the method returns true, it will also return an
+// updated operation filter that includes the operations required by the new
+// registration.
 //
 // NOT safe to call on nil Processor.
 func (p *Processor) Register(
@@ -356,7 +361,7 @@ func (p *Processor) Register(
 	withDiff bool,
 	stream Stream,
 	errC chan<- *roachpb.Error,
-) bool {
+) (bool, *Filter) {
 	// Synchronize the event channel so that this registration doesn't see any
 	// events that were consumed before this registration was called. Instead,
 	// it should see these events during its catch up scan.
@@ -368,9 +373,10 @@ func (p *Processor) Register(
 	)
 	select {
 	case p.regC <- r:
-		return true
+		// Wait for response.
+		return true, <-p.filterResC
 	case <-p.stoppedC:
-		return false
+		return false, nil
 	}
 }
 

--- a/pkg/storage/rangefeed/processor_test.go
+++ b/pkg/storage/rangefeed/processor_test.go
@@ -188,7 +188,7 @@ func TestProcessorBasic(t *testing.T) {
 	// Add a registration.
 	r1Stream := newTestStream()
 	r1ErrC := make(chan *roachpb.Error, 1)
-	r1OK := p.Register(
+	r1OK, r1Filter := p.Register(
 		roachpb.RSpan{Key: roachpb.RKey("a"), EndKey: roachpb.RKey("m")},
 		hlc.Timestamp{WallTime: 1},
 		nil,   /* catchUpIter */
@@ -206,6 +206,14 @@ func TestProcessorBasic(t *testing.T) {
 		)},
 		r1Stream.Events(),
 	)
+
+	// Test the processor's operation filter.
+	require.True(t, r1Filter.NeedVal(roachpb.Span{Key: roachpb.Key("a")}))
+	require.True(t, r1Filter.NeedVal(roachpb.Span{Key: roachpb.Key("d"), EndKey: roachpb.Key("r")}))
+	require.False(t, r1Filter.NeedVal(roachpb.Span{Key: roachpb.Key("z")}))
+	require.False(t, r1Filter.NeedPrevVal(roachpb.Span{Key: roachpb.Key("a")}))
+	require.False(t, r1Filter.NeedPrevVal(roachpb.Span{Key: roachpb.Key("d"), EndKey: roachpb.Key("r")}))
+	require.False(t, r1Filter.NeedPrevVal(roachpb.Span{Key: roachpb.Key("z")}))
 
 	// Test checkpoint with one registration.
 	p.ForwardClosedTS(hlc.Timestamp{WallTime: 5})
@@ -301,14 +309,14 @@ func TestProcessorBasic(t *testing.T) {
 		r1Stream.Events(),
 	)
 
-	// Add another registration.
+	// Add another registration with withDiff = true.
 	r2Stream := newTestStream()
 	r2ErrC := make(chan *roachpb.Error, 1)
-	r2OK := p.Register(
+	r2OK, r1And2Filter := p.Register(
 		roachpb.RSpan{Key: roachpb.RKey("c"), EndKey: roachpb.RKey("z")},
 		hlc.Timestamp{WallTime: 1},
-		nil,   /* catchUpIter */
-		false, /* withDiff */
+		nil,  /* catchUpIter */
+		true, /* withDiff */
 		r2Stream,
 		r2ErrC,
 	)
@@ -322,6 +330,16 @@ func TestProcessorBasic(t *testing.T) {
 		)},
 		r2Stream.Events(),
 	)
+
+	// Test the processor's new operation filter.
+	require.True(t, r1And2Filter.NeedVal(roachpb.Span{Key: roachpb.Key("a")}))
+	require.True(t, r1And2Filter.NeedVal(roachpb.Span{Key: roachpb.Key("y")}))
+	require.True(t, r1And2Filter.NeedVal(roachpb.Span{Key: roachpb.Key("y"), EndKey: roachpb.Key("zzz")}))
+	require.False(t, r1And2Filter.NeedVal(roachpb.Span{Key: roachpb.Key("zzz")}))
+	require.False(t, r1And2Filter.NeedPrevVal(roachpb.Span{Key: roachpb.Key("a")}))
+	require.True(t, r1And2Filter.NeedPrevVal(roachpb.Span{Key: roachpb.Key("y")}))
+	require.True(t, r1And2Filter.NeedPrevVal(roachpb.Span{Key: roachpb.Key("y"), EndKey: roachpb.Key("zzz")}))
+	require.False(t, r1And2Filter.NeedPrevVal(roachpb.Span{Key: roachpb.Key("zzz")}))
 
 	// Both registrations should see checkpoint.
 	p.ForwardClosedTS(hlc.Timestamp{WallTime: 20})
@@ -375,7 +393,7 @@ func TestProcessorBasic(t *testing.T) {
 	// Adding another registration should fail.
 	r3Stream := newTestStream()
 	r3ErrC := make(chan *roachpb.Error, 1)
-	r3OK := p.Register(
+	r3OK, _ := p.Register(
 		roachpb.RSpan{Key: roachpb.RKey("c"), EndKey: roachpb.RKey("z")},
 		hlc.Timestamp{WallTime: 1},
 		nil,   /* catchUpIter */

--- a/pkg/storage/replica_rangefeed.go
+++ b/pkg/storage/replica_rangefeed.go
@@ -254,11 +254,6 @@ func (r *Replica) setRangefeedProcessor(p *rangefeed.Processor) {
 	r.rangefeedMu.Lock()
 	defer r.rangefeedMu.Unlock()
 	r.rangefeedMu.proc = p
-	if !r.updateRangefeedFilterLocked() {
-		// This can't happen. We just set the processor and haven't released the
-		// exclusive lock, so no other goroutine could have stopped it.
-		panic("rangefeed processor unexpectedly stopped")
-	}
 	r.store.addReplicaWithRangefeed(r.RangeID)
 }
 
@@ -278,12 +273,22 @@ func (r *Replica) unsetRangefeedProcessor(p *rangefeed.Processor) {
 	r.unsetRangefeedProcessorLocked(p)
 }
 
+func (r *Replica) setRangefeedFilterLocked(f *rangefeed.Filter) {
+	if f == nil {
+		panic("filter nil")
+	}
+	r.rangefeedMu.opFilter = f
+}
+
 func (r *Replica) updateRangefeedFilterLocked() bool {
-	r.rangefeedMu.opFilter = r.rangefeedMu.proc.Filter()
+	f := r.rangefeedMu.proc.Filter()
 	// Return whether the update to the filter was successful or not. If
 	// the processor was already stopped then we can't update the filter.
-	stopped := r.rangefeedMu.opFilter == nil
-	return !stopped
+	if f != nil {
+		r.setRangefeedFilterLocked(f)
+		return true
+	}
+	return false
 }
 
 // The size of an event is 112 bytes, so this will result in an allocation on
@@ -314,11 +319,12 @@ func (r *Replica) registerWithRangefeedRaftMuLocked(
 	r.rangefeedMu.Lock()
 	p := r.rangefeedMu.proc
 	if p != nil {
-		reg := p.Register(span, startTS, catchupIter, withDiff, stream, errC)
-		if reg && r.updateRangefeedFilterLocked() {
+		reg, filter := p.Register(span, startTS, catchupIter, withDiff, stream, errC)
+		if reg {
 			// Registered successfully with an existing processor.
 			// Update the rangefeed filter to avoid filtering ops
 			// that this new registration might be interested in.
+			r.setRangefeedFilterLocked(filter)
 			r.rangefeedMu.Unlock()
 			return p
 		}
@@ -362,7 +368,7 @@ func (r *Replica) registerWithRangefeedRaftMuLocked(
 	// any other goroutines are able to stop the processor. In other words,
 	// this ensures that the only time the registration fails is during
 	// server shutdown.
-	reg := p.Register(span, startTS, catchupIter, withDiff, stream, errC)
+	reg, filter := p.Register(span, startTS, catchupIter, withDiff, stream, errC)
 	if !reg {
 		catchupIter.Close() // clean up
 		select {
@@ -374,10 +380,11 @@ func (r *Replica) registerWithRangefeedRaftMuLocked(
 		}
 	}
 
-	// Set the rangefeed reference. We know that no other registration
-	// process could have raced with ours because calling this method
-	// requires raftMu to be exclusively locked.
+	// Set the rangefeed processor and filter reference. We know that no other
+	// registration process could have raced with ours because calling this
+	// method requires raftMu to be exclusively locked.
 	r.setRangefeedProcessor(p)
+	r.setRangefeedFilterLocked(filter)
 
 	// Check for an initial closed timestamp update immediately to help
 	// initialize the rangefeed's resolved timestamp as soon as possible.

